### DISCUSSION TITLE
ZACwire sensor support (TSIC temperature)

### DISCUSSION THREAD

**Discussion by blagerweij** [arendst/Tasmota#17900] [Q&A]
Hi,

I've connected a Sonoff Mini R2 to my espresso machine, where the mini will overrule the builtin thermocouplers for brewing and steaming temperature. Since the temperature can reach upto 140 degrees, I cannot use a DS18B20 sensor. Instead I'm using a TSIC 306 , which supports upto 150 degrees celcius. I've developed a small Arduino project to verify this all works, and it does. The idea is that the Sonoff Mini can keep the kettle at exactly 93 degrees.

For the ZACwire protocol I'm using this library: https://github.com/lebuni/ZACwire-Library
However, I would prefer to run Tasmato firmwire, where I can use all the features provided out of the box. The sensor is not yet supported by Tasmota, but I'm more than willing to add support for it. I just need a bit of guidance on the required steps.

Do I just add a .ino file in the tasmota_xsns_sensor (with a unique id), and somehow magically this the sensor will be included? seems almost too good to be true.

I would really appreciate a few pointers on how to add a new sensor, and build a new binary. Since the Sonoff mini does not have a lot of space, I will probably have to customize the included sensors / features.

Thanks in advance,
Barry

**Comment by barbudor** [comment]
A little bit more spells and incantation are required for the magic to work

If you need a lib, it has to be added in the lib folder in one of the lib packages. Most probably the 'div' package.
Be sure the library.properties allows all type of processors (*) see others for example

If you need some gpios, you will need to add them at the end of 2 lists.
There's a doc here https://tasmota.github.io/docs/For-Developers/ but it miss quite some details

Look at other drivers. If it's similar to a Ds18x20, you can use it as example although this one as a lot #ifdef for various options

I'm on my mobile now, so sorry for not giving you more details but tomorrow I can be more accurate if needed

**Reply by blagerweij** [reply]
The sensor is a digital one, uses one GPIO (currently GPIO-04). I've added the library to the lib folder, is there anything I need to do to have the sensor show up in the list of peripherals? Can I just assign a new number (e.g. xsnr-104-zacwire) ? And how can we tell Tasmota that this sensor is a temperature sensor? And finally: currently the Sonoff mini uses 'tasmota-lite', which does have any sensors. I managed to solder the data-line to one of these teeny pads on the mini, so now I need to tweak the image for the Sonoff to include my sensor. How do I go about this?

Thanks for the assistance!
Barry

**Reply by barbudor** [reply]
Hi,
Back home so I now can quote more easily whant you need to do

So you need a GPIO.
I don't know is you expect to be able to support multiple sensors and if they can share the same GPIO like the DS18x20. But let's start 1 GPIO, 1 sensor.
Assuming you want to name the GPIO by the sensor GPIO_TSIC306 and you want it to appear in the menus as "TSIC306" (same if you prefer to call it GPIO_ZAC)

You need to do all the following to add that GPIO to TASMOTA
1) In __every__ language file in the `language/` folder you need to add `#define D_SENSOR_TSIC306 "TSIC306"` which is the string that will be shown in the droplist menu. Start by one but you will have to do it for the PR.
2) In `tasmota_template.h` 
- Add `GPIO_TSIC306` at the end of the table `UserSelectablePins`right before `GPIO_SENSOR_END` (line 209 today)
- Add `D_SENSOR_TSIC306` at the end of `kSensorNames` (line 462) respcting hte synaxt with the `"|"` separator
- Add ` AGPIO(GPIO_TSIC306)` in the `kGpioNiceList` list. Location is not critical but it should fit in the "Single wire sensors" section around lines 686-701
3) In you driver, test that the pin is used in the template/module wiht
```
if (PinUsed(GPIO_TSIC306)) {
   // here you initialize your driver
  // preferably allocate memory dynamically with `calloc()` and not use any global variables (except the pointer to your data)
}
```

Hope this help

If you need deeper review, please publish your fork on github so we can review and comment
